heroku_connect/models.py

```diff
@@ -0,0 +1,342 @@
+
+
+class TriggerLogAbstract(models.Model):
+    """Support for accessing the Heroku Connect Trigger Log data and related actions.
+
+    Heroku Connect uses a Trigger Log table to track local changes to connected models (that is,
+    in the Heroku database. Such changes are recorded as rows in the trigger log and, for
+    read-write mappings, eventually written back to Salesforce.
+
+    Old logs are moved to an archive table (usually after 24 hours), from where they are purged
+    eventually (currently 30 days for paid plans, 7 days for demo). Recent logs are modeled by
+    :class:`TriggerLog`; archived logs by :class:`TriggerLogArchive`.
+
+    The data represented by these models is maintained entirely by Heroku Connect, and is
+    instrumental to its operations; it should therefore not be modified. A possible exception is
+    the ``state`` field, which may be changed as detailed in the `error handling`_ section in the
+    Heroku Connect documentation.
+
+    .. seealso::
+        - :class:`.TriggerLogQuerySet`
+        - `Trigger Log in Heroku Connect docs`_
+
+    .. _Trigger Log in Heroku Connect docs:
+        https://devcenter.heroku.com/articles/writing-data-to-salesforce-with-heroku-connect#understanding-the-trigger-log
+    .. _error handling:
+        https://devcenter.heroku.com/articles/writing-data-to-salesforce-with-heroku-connect#write-errors
+
+    """
+
+    # read-only fields
+    id = models.BigIntegerField(primary_key=True, editable=False)
+    created_at = models.DateTimeField(editable=False, null=True)
+    updated_at = models.DateTimeField(editable=False, null=True)
+    processed_at = models.DateTimeField(editable=False, null=True)
+    table_name = models.CharField(max_length=128, editable=False)
+    record_id = models.BigIntegerField(editable=False)
+    sf_id = models.CharField(max_length=18, editable=False, null=True, db_column='sfid')
+    action = models.CharField(max_length=7, editable=False, choices=TriggerLogAction.choices())
+    sf_message = models.TextField(editable=False, null=True, blank=True)
+
+    # TODO: these are more useful as HStoreFields (if 'django.contrib.postgres' in INSTALLED_APPS)
+    # leave them as textfields for now
+    _values = models.TextField(editable=False, null=True, blank=True, db_column='values')
+    _old = models.TextField(editable=False, null=True, blank=True, db_column='old')
+
+    # editable fields
+    state = models.CharField(max_length=8, null=False, blank=False,
+                             choices=TriggerLogState.choices())
+
+    objects = models.Manager.from_queryset(TriggerLogQuerySet)()
+
+    class Meta:
+        abstract = True
+        managed = False
+        get_latest_by = 'created_at'
+        ordering = ('id',)
+
+    is_archived = False
+
+    @classmethod
+    def capture_insert_from_model(cls, table_name, record_id, *, exclude_fields=()):
+        """Create a fresh insert record from the current model state in the database.
```

Comment:
I prefer if multi line docstrings start with a newline. PEP257 doesn't define that, but it just looks nicer, since all lines start with the same base indentation. Makes it easer for the eyes to read it.